src/cli/cmd/fetch.rs

```diff
@@ -0,0 +1,218 @@
+use std::path::{Path, PathBuf};
+use std::process::ExitCode;
+
+use clap::Parser;
+use color_eyre::eyre::{self, WrapErr as _};
+use color_eyre::Result;
+use tokio::fs;
+use tokio::process::Command;
+use tokio::{fs::OpenOptions, io::AsyncWriteExt};
+
+use crate::cli::cmd::nix_command;
+
+use super::{CommandExecute, FlakeHubClient};
+
+/// First line of the error message printed by Nix when --out-link isn't
+/// supported. We use this as a feature test to determine which copying we do.
+const OUT_LINK_NOT_SUPPORTED: &[u8] = b"error: unrecognised flag '--out-link'";
+
+#[derive(Parser)]
+pub(crate) struct FetchSubcommand {
+    /// The FlakeHub flake reference to fetch.
+    /// References must be of this form: {org}/{flake}/{version_req}#{attr_path}
+    flake_ref: String,
+
+    /// Output link to store paths in, a la Nix's `--out-link` option.
+    #[clap(long)]
+    out_link: Option<String>,
+
+    #[clap(from_global)]
+    api_addr: url::Url,
+
+    #[clap(from_global)]
+    cache_addr: url::Url,
+
+    #[clap(from_global)]
+    frontend_addr: url::Url,
+}
+
+#[async_trait::async_trait]
+impl CommandExecute for FetchSubcommand {
+    #[tracing::instrument(skip_all)]
+    async fn execute(self) -> Result<ExitCode> {
+        let parsed = super::parse_flake_output_ref(&self.frontend_addr, &self.flake_ref)?;
+
+        let resolved_path = FlakeHubClient::resolve(
+            self.api_addr.as_str(),
+            &parsed,
+            /* use scoped token */ true,
+        )
+        .await?;
+
+        tracing::info!(
+            "Resolved {} to {}",
+            self.flake_ref,
+            resolved_path.store_path
+        );
+
+        let token = match resolved_path.token {
+            Some(token) => token,
+            None => eyre::bail!("Did not receive a scoped token from FlakeHub!"),
+        };
+
+        let dir = tempfile::tempdir()?;
+
+        let cache_host = self
+            .cache_addr
+            .host_str()
+            .expect("malformed URL: missing host");
+
+        let netrc_path = create_temp_netrc(dir.path(), cache_host, &token).await?;
+        let token_path = netrc_path.display().to_string();
+
+        let out_link = self.out_link.as_ref().map(String::as_str);
+
+        copy(
+            self.cache_addr.as_str(),
+            &resolved_path.store_path,
+            token_path,
+            out_link,
+        )
+        .await?;
+
+        dir.close()?;
+
+        Ok(ExitCode::SUCCESS)
+    }
+}
+
+async fn create_temp_netrc(dir: &Path, host: &str, token: &str) -> Result<PathBuf> {
```

Comment:
We write this file out elsewhere, can we deduplicate the code for it reasonably?